### ISSUE TITLE
Use upstream MIO or import it more fully

### ISSUE THREAD

**Issue by b4n** [universal-ctags/ctags#1044]
[Upstream MIO](https://github.com/b4n/mio) is somewhat different than what was imported in #870.  The difference was apparently due to @techee trying to make the new library as little intrusive as possible.

However, upstream MIO has some interesting things, most of all: tests.  Especially, it checks that both file and memory implementations behave the same, which is important as it's supposed to be a drop-in replacement of the C file API, so should match it as closely as possible.

Some of the differences:
* test suite
* split backends (it was initially implemented as `if`/`else` like it is here, but I changed that to a vtable later on for maintainability and because profiling didn't show any relevant slowdow)

MIO by itself probably isn't really used (I don't know any packaging of it, I don't know any user of it beside Geany's ctags and now U-CTags), so it might be just as fine to make the copy in here the true upstream -- especially as it seems you started extending it for some specific usage.  My point is that it should at least benefit from test coverage.

I suggest to:
* restore the backend implementation split
* import the test suite

If you think it's a good idea, I'll give it a try.

S/A https://github.com/geany/geany/pull/1070#issuecomment-225580487 and following comments.

**Comment by b4n** [comment]
Of course, another solution is to use the upstream MIO as a normal dependency and leave it external, contributing to it as necessary (and extending locally things that don't make sense to have upstream)

**Comment by codebrainz** [comment]
I think @b4n should choose how it's used/maintained, being the original author, maintainer of the first user (Geany), and contributor to the 2nd project to use it.

**Comment by codebrainz** [comment]
> split backends (it was initially implemented as if/else like it is here, but I changed that to a vtable later on for maintainability and because profiling didn't show any relevant slowdow)

With this vtable scheme, it would simplify porting it to other frameworks, like `fmemopen()` on POSIX, or the equivalent on Win32, while keeping a single stable and usable interface.

**Comment by b4n** [comment]
@codebrainz AFAIK `fmemopen()` can't grow the buffer, and `open_memstream()` doesn't allow for an initial value, so it's a bit trickier to use just as if it was a normal file.  But yeah, `fmemopen()` might be usable for read-only MIO.
But none seem to be a total replacement of the current MIO memory backend.

**Comment by masatake** [comment]
@b4n, thank you for offering.

I agree using the upstream code you maintained is better.
However, as you will see, I exntended local mio copy randomly for ctags specific requirement.
I wonder it is acceptable or not.

I also afraid the integration of build script for windows.




BTW, I'm looking for the way to add a function for "file truncation" to MIO. It is related to #994.
For the memory stream backend, it is easy. For the FILE backend, I cannot find the way other than "reopen".
